src/goto-programs/remove_const_function_pointers.cpp

```diff
@@ -0,0 +1,1012 @@
+          - assigned to a value in an array of functions
+          - assigned to a const struct component
+          Or variations within.
+
+\*******************************************************************/
+
+bool remove_const_function_pointerst::operator()(
+  functionst &out_functions)
+{
+  // Replace all const symbols with there values
+  exprt non_symbol_expression=replace_const_symbols(original_expression);
+  return try_resolve_function_call(non_symbol_expression, out_functions);
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::replace_const_symbols
+
+  Inputs:
+   expression - The expression to resolve symbols in
+
+ Outputs: Returns the a modified version of the expression, with all
+          const symbols resovled to their actual values.
+
+ Purpose: To collapse the symbols down to their values where possible
+          This takes a very general approach, recreating the expr tree
+          exactly as it was and ignoring what type of expressions are found
+          and instead recursses over all the operands.
+
+\*******************************************************************/
+
+exprt remove_const_function_pointerst::replace_const_symbols(
+  const exprt &expression) const
+{
+  if(expression.id()==ID_symbol)
+  {
+    if(is_const_expression(expression))
+    {
+      const symbolt &symbol=
+        symbol_table.lookup(expression.get(ID_identifier));
+      if(symbol.type.id()!=ID_code)
+      {
+        const exprt &symbol_value=symbol.value;
+        return replace_const_symbols(symbol_value);
+      }
+      else
+      {
+        return expression;
+      }
+    }
+    else
+    {
+      return expression;
+    }
+  }
+  else
+  {
+    exprt const_symbol_cleared_expr=expression;
+    const_symbol_cleared_expr.operands().clear();
+    for(const exprt &op : expression.operands())
+    {
+      exprt const_symbol_cleared_op=replace_const_symbols(op);
+      const_symbol_cleared_expr.operands().push_back(const_symbol_cleared_op);
+    }
+
+    return const_symbol_cleared_expr;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::resolve_symbol
+
+  Inputs:
+   symbol_expr - The symbol expression
+
+ Outputs: The expression value of the symbol.
+
+ Purpose: Look up a symbol in the symbol table and return its value
+
+\*******************************************************************/
+
+exprt remove_const_function_pointerst::resolve_symbol(
+  const symbol_exprt &symbol_expr) const
+{
+  const symbolt &symbol=
+    symbol_table.lookup(symbol_expr.get_identifier());
+  return symbol.value;
+}
+
+/*******************************************************************\
+
+Function: remove_const_function_pointerst::try_resolve_function_call
+
+  Inputs:
+   expr - The expression to get the possible function calls
+   out_functions - The functions this expression could be
```

Comment:
I don't think so - the expression is what is being called rather than doing the calling. I've replaced it with 
> the functions this expression could be resolved to

Is that clearer?